src/sage/arith/misc.py

```diff
@@ -1842,76 +1865,29 @@ def gcd(a, b=None, **kwargs):
     """
     # Most common use case first:
     if b is not None:
-        try:
-            m = a.gcd
-        except (AttributeError, TypeError):
-            e = py_scalar_to_element(a)
-            if e is a:
-                raise TypeError("unable to call gcd with {!r}".format(a))
-            m = e.gcd
-
-        try:
-            return m(b, **kwargs)
-        except TypeError:
-            return m(py_scalar_to_element(b), **kwargs)
+        a = py_scalar_to_element(a)
+        b = py_scalar_to_element(b)
+        return a.gcd(b, **kwargs)
 
+    from sage.categories.objects import Objects
     from sage.structure.sequence import Sequence
     seq = Sequence(py_scalar_to_element(el) for el in a)
-    if seq.universe() is ZZ:
-        return GCD_list(seq)
-    else:
-        return __GCD_sequence(seq, **kwargs)
-
+    U = seq.universe()
 
-GCD = gcd
+    if len(seq) == 0:
+        return U(Integer(0))
 
+    if U is Objects():
+        raise ValueError("gcd arguments must share a common parent")
 
-def __GCD_sequence(v, **kwargs):
-    """
-    Internal function returning the gcd of the elements of a sequence
-
-    INPUT:
-
-
-    -  ``v`` - A sequence (possibly empty)
-
-
-    OUTPUT: The gcd of the elements of the sequence as an element of
-    the sequence's universe, or the integer 0 if the sequence is
-    empty.
-
-    EXAMPLES::
+    if U is ZZ:
+        return GCD_list(seq)
 
-        sage: from sage.arith.misc import __GCD_sequence
-        sage: from sage.structure.sequence import Sequence
-        sage: l = ()
-        sage: __GCD_sequence(l)
-        0
-        sage: __GCD_sequence(Sequence(srange(10)))
-        1
-        sage: X=polygen(QQ)
-        sage: __GCD_sequence(Sequence((2*X+4,2*X^2,2)))
-        1
-        sage: X=polygen(ZZ)
-        sage: __GCD_sequence(Sequence((2*X+4,2*X^2,2)))
-        2
-        sage: __GCD_sequence(Sequence((1/1,1/2)))
-        1/2
+    from functools import reduce
+    return reduce(gcd, map(U, seq))
```

Comment:
```suggestion
    from functools import reduce
    return reduce(gcd, map(U, seq))
```
This should be more expensive than the previous implementation as it goes through all of the checks with additional redirections. Please check the timings as this is an important function. (Although use `ZZ.zero()` instead of `ZZ(0)` since the former is cached.)

Reply:
Oh wow, I didn't realise how much slower this is. Comparing `reduce(gcd, arr)` with the explicit for loop on `x = randrange(2**32); arr = [ZZ(x * randrange(2**32)) for _ in range(10**6)]` shows reduce is 40% to 50% slower. Sorry for that